04-onion-routing.md

```diff
@@ -263,39 +264,105 @@ It is formatted according to the Type-Length-Value format defined in [BOLT #1](0
     2. data:
         * [`32*byte`:`payment_secret`]
         * [`tu64`:`total_msat`]
+    1. type: 10 (`encrypted_recipient_data`)
+    2. data:
+        * [`...*byte`:`encrypted_data`]
+    1. type: 12 (`current_blinding_point`)
+    2. data:
+        * [`point`:`blinding`]
     1. type: 16 (`payment_metadata`)
     2. data:
         * [`...*byte`:`payment_metadata`]
+    1. type: 18 (`total_amount_msat`)
+    2. data:
+        * [`tu64`:`total_msat`]
 
 ### Requirements
 
-The writer:
+The creator of `encrypted_recipient_data` (usually, the recipient of payment):
+
+  - MUST create `encrypted_data_tlv` for each node in the blinded route (including itself).
+  - MUST include `encrypted_data_tlv.short_channel_id` and `encrypted_data_tlv.payment_relay` for each non-final node.
+  - SHOULD set `encrypted_data_tlv.payment_constraints` and `encrypted_data_tlv.allowed_features` for each node in the blinded route (including itself):
+    - `max_cltv_expiry` to the largest block height at which the route is allowed to be used, starting
+    from the final node and adding `encrypted_data_tlv.payment_relay.cltv_expiry_delta` at each hop.
+    - `htlc_minimum_msat` to the largest minimum HTLC value the nodes will allow.
+    - `allowed_features.features` to be empty.
```

Comment:
Hmm, I don't think we should set allowed_features, and if we do MUST set it to empty?

Reply:
I like that it's really explicit if we set it with an empty features array: the reader sees an `allowed_features` field containing no features, so it knows this payment cannot use any fancy feature. It's also fine if it's not set, in that case it should be interpreted as if it were set and empty. I updated the wording in https://github.com/lightning/bolts/pull/765/commits/a8778d10d7bc37801da6255987ada3168aa2e05f

I think it's also important that it's set in the test vector, to ensure implementers have taken this field into account even though we don't yet have features impacting this, otherwise there would be a risk that implementers would simply forget about implementing this check.